rfcs/inline_symbols.md

```diff
@@ -0,0 +1,508 @@
+`symbol.null` using type descriptor `0xF3`.
+
+```
+        7       4 3       0
+        +---------+---------+
+        |   15    |    3    |
+        +---------+---------+======+
+        |     Length [VarUInt]     |
+        +--------------------------+
+        |     UTF8 Representation  |
+        +--------------------------+
+```
+
+While many applications simply treat strings and symbols as different encoding options for text,
+others rely on the distinction between the types for their business logic. Providing this encoding
+ensures that a symbol value can be written as cheaply as a string value in any use case.
+
+#### `0xF4`: Inline symbol structs
+
+Inline symbol structs are structs whose field names can be encoded as either symbol IDs or
+inline symbols. The encoding need not be homogenous; some fields can use symbol IDs while
+others use inline symbols.
+
+Inline symbol structs have a type descriptor byte of `0xF4`. A `Length` field containing the number
+of bytes in the struct's representation must always follow the type descriptor byte.
+
+Inline symbol structs' field names are encoded as a `VarInt` (not a `VarUInt`). The sign bit is used
+to indicate whether the field name has been encoded as an inline symbol or as a symbol ID. 
+
+
+```
+        7       4 3       0
+        +---------+---------+
+        |   15    |    4    |
+        +---------+---------+======+
+        :     length [VarUInt]     :
+        +==========================+=================================================+
+        : field name [VarInt + optional UTF8 representation]  :        value         :
+        +============================================================================+
+                    ⋮                     ⋮
+```
+
+If the `VarInt` is positive, then its magnitude represents a symbol ID. Its text can be found in
+the active symbol table.
+
+If the `VarInt` is negative, then its magnitude represents the length of the field name's UTF8
+representation, which follows immediately after the `Length`.
+
+For example, this struct:
+```js
+{
+    $37: 5,
+    foo: 9
+} 
+```
+could be encoded as:
+```
+                   f  o  o
+F4 89 A5 21 05 C3 66 6f 6f 21 09
+ ^  ^  ^  ^  ^  ^           ^  ^---- Value for "foo": 9
+ |  |  |  |  |  |           +------- 1-byte positive integer
+ |  |  |  |  |  +------------------- 0b1100_0011 / VarInt -3 / a 3-byte inline symbol field name ("foo")
+ |  |  |  |  +---------------------- Value for $37: 5
+ |  |  |  +------------------------- 1-byte positive integer
+ |  |  +---------------------------- 0b1010_0101 / VarInt 37 / symbol ID field name ($37)
+ |  +------------------------------- VarUInt Length: 9 bytes
+ +---------------------------------- Type descriptor: inline symbol struct
+```
+
+If the sign bit is positive and the magnitude is zero, the field name is the empty string.
+
+If the sign bit is negative and the magnitude is zero, the field name is symbol ID `0`.
```

Comment:
Should these be switched around? Sign bit positive means the magnitude is a symbol ID, so it would make sense for positive + 0 to mean SID 0 and negative + 0 to mean empty string.

Reply:
Good catch, this is a holdover from an earlier draft that inverted the meaning of the sign bit. Fixed.